docs/_posts/2020-02-28-puppetwash.md

```diff
@@ -0,0 +1,382 @@
+│   ├── netmask
+│   ├── netmask_eth0
+│   ├── netmask_lo
+│   ├── network
+│   ├── network_eth0
+│   ├── network_lo
+│   ├── networking
+│   ├── operatingsystem
+│   ├── operatingsystemmajrelease
+│   ├── operatingsystemrelease
+│   ├── os
+│   ├── osfamily
+│   ├── partitions
+│   ├── path
+│   ├── physicalprocessorcount
+│   ├── processor0
+│   ├── processor1
+│   ├── processor2
+│   ├── processor3
+│   ├── processor4
+│   ├── processor5
+│   ├── processorcount
+│   ├── processors
+│   ├── productname
+│   ├── puppetversion
+│   ├── ruby
+│   ├── rubyplatform
+│   ├── rubysitedir
+│   ├── rubyversion
+│   ├── selinux
+│   ├── serialnumber
+│   ├── swapfree
+│   ├── swapfree_mb
+│   ├── swapsize
+│   ├── swapsize_mb
+│   ├── system_uptime
+│   ├── timezone
+│   ├── trusted
+│   ├── uptime
+│   ├── uptime_days
+│   ├── uptime_hours
+│   ├── uptime_seconds
+│   ├── uuid
+│   └── virtual
+└── reports
+    └── 2020-02-28T00:08:28.495Z
+
+2 directories, 107 files
+```
+
+Pupperware presents the last catalog
+```
+wash puppet/pupperware/nodes > cd puppet.test
+wash puppet/pupperware/nodes/puppet.test > head catalog.json
+{
+  "catalog_uuid": "bf1576bd-8406-4a9b-ba9d-9adde461f732",
+  "producer": "puppet.test",
+  "hash": "6ddc338fa1582e9b5b58b3c38ff307dca4a1269a",
+  "transaction_uuid": "510c3b1a-967f-4b88-a66d-225bc40ac467",
+  "producer_timestamp": "2020-02-28T00:08:28.262Z",
+  "environment": "production",
+  "code_id": null,
+  "version": "1582848507",
+  "resources": {
+```
+reports
+```
+wash puppet/pupperware/nodes/puppet.test > head reports/2020-02-28T00:08:28.495Z
+[
+  {
+    "catalog_uuid": "bf1576bd-8406-4a9b-ba9d-9adde461f732",
+    "receive_time": "2020-02-28T00:08:28.727Z",
+    "producer": "puppet.test",
+    "hash": "8043a61b020be5ed2f161eac6e1d7f1df2ee8fcf",
+    "transaction_uuid": "510c3b1a-967f-4b88-a66d-225bc40ac467",
+    "puppet_version": "6.7.2",
+    "noop": true,
+    "corrective_change": null,
+```
+and individual fact values
+```
+wash puppet/pupperware/nodes/puppet.test > cat facts/osfamily
+Debian
+```
+
+Nodes have metadata attached to them that we can filter on
+```
+wash puppet/pupperware/nodes/puppet.test > meta .
+cached_catalog_status: not_used
+catalog_environment: production
+catalog_timestamp: "2020-02-28T00:08:28.328Z"
+certname: puppet.test
+deactivated: null
+expired: null
+facts_environment: production
+facts_timestamp: "2020-02-28T00:08:27.886Z"
+latest_report_corrective_change: null
+latest_report_hash: 8043a61b020be5ed2f161eac6e1d7f1df2ee8fcf
+latest_report_job_id: null
+latest_report_noop: true
+latest_report_noop_pending: false
+latest_report_status: unchanged
+report_environment: production
+report_timestamp: "2020-02-28T00:08:28.495Z"
+```
+
+For example, we could search for nodes that have been deactivated (or that haven't been)
+```
+wash puppet/pupperware/nodes/puppet.test > find .. -meta deactivated -exists
+wash puppet/pupperware/nodes/puppet.test > find .. -meta deactivated -null
```

Comment:
Why not just do this in the `nodes` directory? `..` requires some unnecessary thinking on the reader's part to figure out that you're searching for this in the `nodes` directory. Also think `-k '*node` would be useful but that's a nit.

Reply:
Also think the above `meta <node>` example should be in `nodes` directory.